mono/metadata/handle.h

```diff
@@ -338,23 +406,81 @@ Handle macros/functions
 #define MONO_HANDLE_PAYLOAD_OFFSET_(PayloadType) MONO_STRUCT_OFFSET(PayloadType, __raw)
 #define MONO_HANDLE_PAYLOAD_OFFSET(TYPE) MONO_HANDLE_PAYLOAD_OFFSET_(TYPED_HANDLE_PAYLOAD_NAME (TYPE))
 
-#define MONO_HANDLE_INIT ((void*) mono_null_value_handle)
 #define NULL_HANDLE mono_null_value_handle
 
 //XXX add functions to get/set raw, set field, set field to null, set array, set array to null
-#define MONO_HANDLE_RAW(HANDLE) ((HANDLE)->__raw)
 #define MONO_HANDLE_DCL(TYPE, NAME) TYPED_HANDLE_NAME(TYPE) NAME = MONO_HANDLE_NEW (TYPE, (NAME ## _raw))
 
+#if defined (__GNUC__) && !defined (__clang__)
+static inline MONO_ALWAYS_INLINE gpointer
+mono_remove_function_type_to_suppress_gcc_warning (gpointer a)
+// gcc warning: function called through a non-compatible type
+// Passing the function through this function suppresses the warning.
+{
+	return a;
+}
+#else
+#define mono_remove_function_type_to_suppress_gcc_warning(a) (a)
+#endif
+
+#if MONO_TYPE_SAFE_HANDLES
+
 #ifndef MONO_HANDLE_TRACK_OWNER
-#define MONO_HANDLE_NEW(TYPE, VALUE) (TYPED_HANDLE_NAME(TYPE))( mono_handle_new ((MonoObject*)(VALUE)) )
+
+// Structs cannot be cast to structs.
+// Therefore, cast the function pointer to change its return type and parameter type.
+#define MONO_HANDLE_NEW(type, object) \
+	(((TYPED_HANDLE_NAME (type)(*) (type*)) \
+		mono_remove_function_type_to_suppress_gcc_warning (mono_handle_new)) (object))
+
+#else
+
+// Structs cannot be cast to structs.
+// Therefore, cast the function pointer to change its return type and parameter type.
+#define MONO_HANDLE_NEW(type, object) \
+	(((TYPED_HANDLE_NAME (type)(*) (type*, const char*)) \
+		mono_remove_function_type_to_suppress_gcc_warning (mono_handle_new)) ((object), HANDLE_OWNER))
+
+#endif
+
+// Structs cannot be cast to structs.
+// Therefore, cast the function pointer to change its return type.
+#define MONO_HANDLE_CAST(type, value) \
+	(((TYPED_HANDLE_NAME (type) (*) (gpointer)) \
+		mono_remove_function_type_to_suppress_gcc_warning (mono_handle_cast)) ((value).__raw))
+
+// With Visual C++ compiling as C, the type of a ternary expression
+// yielding two unrelated non-void pointers is the type of the first, plus a warning.
+// This can be used to simulate gcc typeof extension.
```

Comment:
Wow.

Do you think it's worth making this a more general utility somewhere in `utils/` and calling it here?  Might make the definition of `MONO_HANDLE_RAW` a bit cleaner.

Reply:
I'm uneasy about this. I didn't find it in documentation. Documentation just says ternary does complicated type manipulation and discourages its  use. I was trying to think of how to use ternary to "preserve" a type in one place and apply it another and discovered it. I was hoping void* and T* would yield T*, but it yields void*. So I tried T1* and T2*. Turning off type-safety of handles when using Visual C++ is another choice. And C++ should allow doing this properly.

Or maybe it is wonderful??

Any on/off of type-safety will mean some builds succeed where others do not which is always unfortunate.

Reply:
This is required though to have MONO_HANDLE_RAW both support NULL_HANDLE, and not  double evaluate. Historically it null deferences on NULL_HANDLE.

I found this because I was removing the double null checks in  favor of existing MONO_HANDLE_IS_NULL -- which we could require gate all possibly-null MONO_HANDLE_RAWs.